File: server/relay/envelope.py
"""Enveloppe JSON multiplexée entre le relay server et l'app Android.

Format : {version, channel, type, id, payload}. `version` est obligatoire dès
le départ pour permettre de faire évoluer le protocole plus tard sans casser
silencieusement la compatibilité entre app et serveur.
"""

from __future__ import annotations

import uuid
from dataclasses import dataclass, field
from typing import Any

PROTOCOL_VERSION = 1

CHANNELS = ("system", "chat", "proactive", "bridge")


class EnvelopeError(ValueError):
    pass


@dataclass
class Envelope:
    channel: str
    type: str
    payload: dict[str, Any]
    version: int = PROTOCOL_VERSION
    id: str = field(default_factory=lambda: str(uuid.uuid4()))

    def to_dict(self) -> dict[str, Any]:
        return {
            "version": self.version,
            "channel": self.channel,
            "type": self.type,
            "id": self.id,
            "payload": self.payload,
        }

    @staticmethod
    def from_dict(data: dict[str, Any]) -> "Envelope":
        if "version" not in data:
            raise EnvelopeError("missing required field 'version'")
        if not isinstance(data["version"], int):
            raise EnvelopeError("'version' must be an int")
        if data["version"] != PROTOCOL_VERSION:
            raise EnvelopeError(
                f"unsupported protocol version {data['version']!r} "
                f"(expected {PROTOCOL_VERSION})"
            )

        channel = data.get("channel")
        if channel not in CHANNELS:
            raise EnvelopeError(f"invalid or missing channel {channel!r}")

        msg_type = data.get("type")
        if not isinstance(msg_type, str) or not msg_type:
            raise EnvelopeError("missing required field 'type'")

        payload = data.get("payload")
        if not isinstance(payload, dict):
            raise EnvelopeError("missing or invalid 'payload' (must be an object)")

        envelope_id = data.get("id") or str(uuid.uuid4())

        return Envelope(
            channel=channel,
            type=msg_type,
            payload=payload,
            version=data["version"],
            id=envelope_id,
        )

File: app/src/main/java/com/hasan/v1/network/models/Envelope.kt
package com.hasan.v1.network.models

import org.json.JSONObject
import java.util.UUID

/**
 * Enveloppe JSON multiplexée entre l'app et le relay server.
 *
 * Format : {version, channel, type, id, payload}. `version` est obligatoire
 * dès le départ pour permettre de faire évoluer le protocole plus tard sans
 * casser silencieusement la compatibilité entre app et serveur — même
 * contrat que côté serveur (voir server/relay/envelope.py).
 */
data class Envelope(
    val channel: String,
    val type: String,
    val payload: JSONObject,
    val version: Int = PROTOCOL_VERSION,
    val id: String = UUID.randomUUID().toString()
) {
    fun toJson(): JSONObject = JSONObject().apply {
        put("version", version)
        put("channel", channel)
        put("type", type)
        put("id", id)
        put("payload", payload)
    }

    override fun toString(): String = toJson().toString()

    companion object {
        const val PROTOCOL_VERSION = 1

        val CHANNELS = setOf("system", "chat", "proactive", "bridge")

        /** Parse une enveloppe reçue. Retourne null si le format est invalide (log côté appelant). */
        fun fromJson(raw: String): Envelope? {
            return try {
                val obj = JSONObject(raw)
                fromJsonObject(obj)
            } catch (_: Exception) {
                null
            }
        }

        fun fromJsonObject(obj: JSONObject): Envelope? {
            if (!obj.has("version") || obj.optInt("version", -1) != PROTOCOL_VERSION) return null

            val channel = obj.optString("channel", "")
            if (channel !in CHANNELS) return null

            val type = obj.optString("type", "")
            if (type.isBlank()) return null

            val payload = obj.optJSONObject("payload") ?: return null

            val id = obj.optString("id").takeIf { it.isNotBlank() } ?: UUID.randomUUID().toString()

            return Envelope(channel = channel, type = type, payload = payload, version = PROTOCOL_VERSION, id = id)
        }
    }
}
